### PR TITLE
Fix: Material Slider Theme

### PR DESCRIPTION
**Pull request by Amjad50**
## Description

This Pull Request fixed a wrong render in TrackShapes of Material Slider widget.
As shown

#### before
before

#### after
after

## Related Issues

After discovering the issue with the TrackShape, I created an issue in github and was asked to implement a change as a PL.
#37683

## Tests
- edited:
    + Default slider track draws correctly
    + The slider track height can be overridden
    + The default slider thumb shape sizes can be overridden
- new:
    + Default slider edges draw correctly

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by Amjad50** [comment]
for [this](https://cirrus-ci.com/task/4869392302604288) test, it fails on a single test:
- `TextPainter line metrics` from `test/painting/text_painter_test.dart (749:5)`

And it does not have anything to do with my edits.

This test also yields a success on my machine, I'm not really sure what is the problem here.

File: packages/flutter/lib/src/material/slider_theme.dart [hunk 1]
```diff
@@ -1683,18 +1683,46 @@ class RoundedRectSliderTrackShape extends SliderTrackShape with BaseSliderTrackS
     );
 
     // The arc rects create a semi-circle with radius equal to track height.
-    final Rect leftTrackArcRect = Rect.fromLTWH(trackRect.left, trackRect.top, trackRect.height, trackRect.height);
+    final Rect leftTrackArcRect = Rect.fromLTWH(
+      trackRect.left - trackRect.height / 2,
+      trackRect.top,
+      trackRect.height,
+      trackRect.height,
+    );
```
Comment (Amjad50) [inline packages/flutter/lib/src/material/slider_theme.dart]: moved the ARC to the left by height/2, as the thumb should not go to the edge as it was before

File: packages/flutter/lib/src/material/slider_theme.dart [hunk 2]
```diff
@@ -1683,18 +1683,46 @@ class RoundedRectSliderTrackShape extends SliderTrackShape with BaseSliderTrackS
...
+
+    final Rect rightTrackArcRect = Rect.fromLTWH(
+      trackRect.right - trackRect.height / 2,
+      trackRect.top,
+      trackRect.height,
+      trackRect.height,
+    );
```
Comment (Amjad50) [inline packages/flutter/lib/src/material/slider_theme.dart]: not changed, just moved it here for better readability

File: packages/flutter/lib/src/material/slider_theme.dart [hunk 3]
```diff
@@ -1683,18 +1683,46 @@ class RoundedRectSliderTrackShape extends SliderTrackShape with BaseSliderTrackS
...
+
+    final Rect leftTrackSegment = Rect.fromLTRB(
+      trackRect.left,
+      trackRect.top,
+      thumbCenter.dx,
+      trackRect.bottom,
+    );
```
Comment (Amjad50) [inline packages/flutter/lib/src/material/slider_theme.dart]: merged into the thumb, as in before it was only around the thumb. Which created a gap in rounded or not rect thumbs

File: packages/flutter/lib/src/material/slider_theme.dart [hunk 4]
```diff
@@ -1683,18 +1683,46 @@ class RoundedRectSliderTrackShape extends SliderTrackShape with BaseSliderTrackS
...
+
+    final Rect rightTrackSegment = Rect.fromLTRB(
+      thumbCenter.dx,
+      trackRect.top,
+      trackRect.right,
+      trackRect.bottom,
+    );
```
Comment (Amjad50) [inline packages/flutter/lib/src/material/slider_theme.dart]: same leftTrackSegment

File: packages/flutter/lib/src/material/slider_theme.dart [hunk 5]
```diff
@@ -1683,18 +1683,46 @@ class RoundedRectSliderTrackShape extends SliderTrackShape with BaseSliderTrackS
...
     if (!leftTrackArcRect.isEmpty)
-      context.canvas.drawArc(leftTrackArcRect, math.pi / 2, math.pi, false, leftTrackPaint);
-    final Rect rightTrackArcRect = Rect.fromLTWH(trackRect.right - trackRect.height / 2, trackRect.top, trackRect.height, trackRect.height);
+      context.canvas.drawArc(leftTrackArcRect, math.pi / 2, math.pi, false,
+          leftTrackSegment.isEmpty ? rightTrackPaint : leftTrackPaint);
+
     if (!rightTrackArcRect.isEmpty)
-      context.canvas.drawArc(rightTrackArcRect, -math.pi / 2, math.pi, false, rightTrackPaint);
+      context.canvas.drawArc(rightTrackArcRect, -math.pi / 2, math.pi, false,
+          rightTrackSegment.isEmpty ? leftTrackPaint : rightTrackPaint);
```
Comment (Amjad50) [inline packages/flutter/lib/src/material/slider_theme.dart]: filling the right arc if the shape is full, and emptying the left if it is empty.
As it was in before, left is always filled and right is always emptyed.

File: packages/flutter/lib/src/material/slider_theme.dart [hunk 6]
```diff
@@ -1944,23 +1969,55 @@ class RoundedRectRangeSliderTrackShape extends RangeSliderTrackShape {
...
-    final Rect rightTrackSegment = Rect.fromLTRB(rightThumbOffset.dx + thumbRadius, trackRect.top, trackRect.right - trackRadius, trackRect.bottom);
+
     if (!rightTrackSegment.isEmpty)
       context.canvas.drawRect(rightTrackSegment, inactivePaint);
 
-    final Rect rightTrackArcRect = Rect.fromLTWH(trackRect.right - trackRect.height, trackRect.top, trackRect.height, trackRect.height);
     if (!rightTrackArcRect.isEmpty)
-      context.canvas.drawArc(rightTrackArcRect, -math.pi / 2, math.pi, false, inactivePaint);
+      context.canvas.drawArc(
+          rightTrackArcRect, -math.pi / 2, math.pi, false, rightTrackSegment.isEmpty ? activePaint : inactivePaint);
```
Comment (Amjad50) [inline packages/flutter/lib/src/material/slider_theme.dart]: same as the above class changes.\

File: packages/flutter/test/material/slider_theme_test.dart [hunk 7]
```diff
@@ -229,28 +229,71 @@ void main() {
...
-    // the center of the thumb.
+    // the center of the thumb to the furthest points the center of the thumb
+    // can be at, 0 and 1 (empty and full).
     expect(
       sliderBox,
       paints
-        ..rect(rect: const Rect.fromLTRB(25.0, 299.0, 202.0, 301.0), color: sliderTheme.activeTrackColor)
-        ..rect(rect: const Rect.fromLTRB(222.0, 299.0, 776.0, 301.0), color: sliderTheme.inactiveTrackColor),
+        ..rect(rect: const Rect.fromLTRB(24.0, 299.0, 212.0, 301.0), color: sliderTheme.activeTrackColor)
+        ..rect(rect: const Rect.fromLTRB(212.0, 299.0, 776.0, 301.0), color: sliderTheme.inactiveTrackColor),
```
Comment (Amjad50) [inline packages/flutter/test/material/slider_theme_test.dart]: change values to work with my changes

File: packages/flutter/test/material/slider_theme_test.dart [hunk 8]
```diff
@@ -229,28 +229,71 @@ void main() {
...
+    await tester.pumpWidget(_buildApp(sliderTheme, value: 0.5));
+    await tester.pumpAndSettle(); // wait for disable animation
+
+    // when the slider is in between empty and full, one of the edges will
+    // not be active but the other edge will be.
+    expect(
+      sliderBox,
+      paints
+        ..arc(color: sliderTheme.activeTrackColor)
+        ..arc(color: sliderTheme.inactiveTrackColor),
```
Comment (Amjad50) [inline packages/flutter/test/material/slider_theme_test.dart]: tests the change of the left and right arc fill color based on fill percentage

File: packages/flutter/test/material/slider_theme_test.dart [hunk 9]
```diff
@@ -538,23 +581,24 @@ void main() {
...
-    // The disabled thumb is smaller so the active track has to paint longer to
-    // get to the edge.
+    // the same as the above, just a difference in colors
     expect(
       sliderBox,
       paints
-        ..rect(rect: const Rect.fromLTRB(32.0, 292.0, 202.0, 308.0), color: sliderTheme.disabledActiveTrackColor)
-        ..rect(rect: const Rect.fromLTRB(222.0, 292.0, 776.0, 308.0), color: sliderTheme.disabledInactiveTrackColor),
+        ..rect(rect: const Rect.fromLTRB(24.0, 292.0, 212.0, 308.0), color: sliderTheme.disabledActiveTrackColor)
+        ..rect(rect: const Rect.fromLTRB(212.0, 292.0, 776.0, 308.0), color: sliderTheme.disabledInactiveTrackColor),
```
Comment (Amjad50) [inline packages/flutter/test/material/slider_theme_test.dart]: change values to work with my changes

**Comment by HansMuller** [comment]
CC @clocksmith

**Comment by HansMuller** [comment]
It looks like this PR significantly overlaps with an existing Slider PR that's close to landing: https://github.com/flutter/flutter/pull/45911. The visual changes, including changes to slider track rendering, are detailed in https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/flutter-announce/69dmlKUL5Ew/tQh-ajiEAAAJ.

Please take a look at #45911 and let us know if it's missing anything essential that's defined in this PR.

**Comment by Amjad50** [comment]
Yes, I went through #45911. It fixes most of what is in this PR, except for one small part.

- filling of the arcs on the left and right sides of the rounded shaped track depending on the filling of the track segments
``` dart
context.canvas.drawArc(leftTrackArcRect, math.pi / 2, math.pi, false,
          leftTrackSegment.isEmpty ? rightTrackPaint : leftTrackPaint);
```
Everything else is very good

**Comment by Amjad50** [comment]
@HansMuller should I close this PR as its not important anymore?

**Comment by HansMuller** [comment]
@Amjad50 - we appreciate all of the effort that you put into this! Since #45911 has landed, it would be best to close this one.

**Pull request closed without merging**